lib_gb/src/ppu/gb_ppu.rs

```diff
@@ -104,60 +107,59 @@ impl GbPpu {
     }
 
     fn get_ppu_state(cycle_counter:u32, last_ly:u8)->PpuState{
-        if last_ly > SCREEN_HEIGHT as u8{
+        if last_ly >= SCREEN_HEIGHT as u8{
             return PpuState::Vblank;
         }
 
         //getting the reminder of the clocks 
         let current_line_clocks = cycle_counter % DRAWING_CYCLE_CLOCKS as u32;
         
-        const PIXEL_TRANSFER_START:u8 = OAM_CLOCKS+1;
-        const PIXEL_TRANSFER_END:u8 = OAM_CLOCKS + PIXEL_TRANSFER_CLOCKS;
-        const H_BLANK_START:u8 = PIXEL_TRANSFER_END+1;
-        const H_BLANK_END:u8 = PIXEL_TRANSFER_END + H_BLANK_CLOCKS;
+        const OAM_SERACH_END:u8 = OAM_CLOCKS - 1;
+        const PIXEL_TRANSFER_START:u8 = OAM_CLOCKS;
+        const PIXEL_TRANSFER_END:u8 = OAM_CLOCKS + PIXEL_TRANSFER_CLOCKS - 1;
+        const H_BLANK_START:u8 = OAM_CLOCKS + PIXEL_TRANSFER_CLOCKS;
+        const H_BLANK_END:u8 = H_BLANK_START + H_BLANK_CLOCKS - 1;
 
         return match current_line_clocks as u8{
-            0 ..= OAM_CLOCKS => PpuState::OamSearch,
-            PIXEL_TRANSFER_START ..= PIXEL_TRANSFER_END => PpuState::PixelTransfer,
-            H_BLANK_START ..= H_BLANK_END => PpuState::Hblank,
+            0 ..= OAM_SERACH_END => PpuState::OamSearch, // 0-19 (20)
+            PIXEL_TRANSFER_START ..= PIXEL_TRANSFER_END => PpuState::PixelTransfer, //20-62 (43)
+            H_BLANK_START ..= H_BLANK_END => PpuState::Hblank,//63-113(51)
             _=>std::panic!("Error calculating ppu state")
         };
     }
 
-    //receiving the amount of cycles from the first frame. because of this the PPU will be syncronized to the frames of the program.
-    //the down side is that upon turnning the PPU on it will start from the middle of the frame and not from the start.
-    pub fn update_gb_screen(&mut self, memory: &dyn ReadOnlyVideoMemory, cycles_passed:u32){
+    pub fn update_gb_screen(&mut self, memory: &dyn UnprotectedMemory, cycles_passed:u32){
         if !self.screen_enable{
             self.current_line_drawn = 0;
+            self.current_cycle = 0;
             self.screen_buffer = [Self::color_as_uint(&WHITE);SCREEN_HEIGHT * SCREEN_WIDTH];
             self.state = PpuState::Hblank;
             return;
         }
 
-        self.update_ly(cycles_passed);
-        self.state = Self::get_ppu_state(cycles_passed, self.current_line_drawn);
+        self.current_cycle += cycles_passed as u32;
+        self.state = Self::get_ppu_state(self.current_cycle, self.current_line_drawn);
+        self.update_ly();
```

Comment:
Why swapping those lines breaks the acid tests????

Reply:
I really dont know how it worked until now.

When it turn 144 first it will still be in mode 2 (oam search) in the first cycle and than it is catching up

Reply:
Maybe the update ly doesnt ever reaches 154 or more and the timing is screwed

Reply:
it was the fucking stat interrupt